File: app/src/main/java/com/example/games/parejas/CardsGameContract.kt
package com.example.games.parejas

import pro.respawn.flowmvi.api.MVIAction
import pro.respawn.flowmvi.api.MVIIntent
import pro.respawn.flowmvi.api.MVIState

/**
 * Contrato reactivo de "Parejas Ocultas" sobre FlowMVI real (`pro.respawn.flowmvi`).
 * Reemplaza al contrato nativo de la iteración anterior (`ParejasOcultasState`/Intent/
 * Action) ahora que el motor pasó a ser un [pro.respawn.flowmvi.api.Container] de
 * verdad — ver [CardsGameContainer].
 */

data class MemoryCardUi(
  val id: Int,
  val pairKey: String,
  val symbol: String,
  val isFaceUp: Boolean = false,
  val isMatched: Boolean = false
)

enum class SoundEffect { FLIP, MATCH, MISMATCH, LEVEL_UP, ROUND_COMPLETE }
enum class HapticPattern { CORRECT, INCORRECT, LEVEL_UP }
enum class CountdownReason { START, ADVANCE, DEMOTED, RETRY }

sealed interface CardsGameState : MVIState {

  data object Stopped : CardsGameState

  /** Precarga (plugin `asyncCache`/`cache`): construir el mazo + seleccionar el banco
   *  de símbolos según la interferencia perceptual vigente, antes de mostrar nada. */
  data class Loading(val progress: Float) : CardsGameState

  /**
   * Pantalla de "prepárate" entre niveles (y al arrancar la partida). Pedido explícito
   * de Ricardo tras jugarlo: "se ve muy estática" -- al completar un tablero, la partida
   * ya no termina ahí, pasa DE INMEDIATO al siguiente nivel con esta cuenta 3-2-1 como
   * transición, hasta [TOTAL_STAGES] niveles. [reason] distingue arranque/avance normal
   * de una bajada o repetición de nivel por fallas (ver [CardsGameContainer.handleStageFailure]),
   * para que el cambio de nivel hacia ATRÁS se lea como mecánica del juego y no como un bug.
   */
  data class Countdown(
    val secondsLeft: Int,
    val stage: Int,
    val totalStages: Int,
    val reason: CountdownReason = CountdownReason.ADVANCE
  ) : CardsGameState

  data class Running(
    val cards: List<MemoryCardUi>,
    val gridDimensions: Pair<Int, Int>, // (columnas, filas)
    val score: Int,
    val matchedPairs: Int,
    val totalPairs: Int,
    val attempts: Int,
    val isMemorizing: Boolean,
    val memorizeMillisLeft: Long,
    /** Nivel actual dentro de esta partida (1..[TOTAL_STAGES]) -- distinto del "Nivel N"
     *  de dificultad adaptativa elegido en el diálogo de la biblioteca, este es el
     *  contador de tableros sucesivos DENTRO de la sesión. */
    val stage: Int = 1,
    val totalStages: Int = TOTAL_STAGES,
    /** Candado de turno visible en UI. La serialización real de golpes casi
     *  simultáneos la garantiza el propio Store de FlowMVI (procesa un MVIIntent a
     *  la vez desde su canal interno) — este flag es la señal para deshabilitar el
     *  `clickable` de las cartas mientras se resuelve la pareja anterior, no el
     *  mecanismo de seguridad en sí. */
    val isTurnLocked: Boolean,
    val currentStreak: Int = 0,
    val difficultyIndex: Float,
    val distractorCount: Int,
    val distractorOpacity: Float,
    val distractorSeed: Int,
    val timeLeftSeconds: Int? = null,
    /** Presupuesto inicial de [timeLeftSeconds] (Modo Reto) — sin esto la barra de
     *  tiempo de la UI no tiene contra qué calcular el porcentaje restante. */
    val totalTimeSeconds: Int? = null,
    /** Cronómetro de tiempo de resolución (desde que termina la memorización, NO desde
     *  que arranca la ronda) — pedido explícito de Ricardo tras jugarlo: "el tiempo en
     *  las actividades también es una métrica relevante", incluso en Precisión (sin
     *  reloj). Alimenta la barra de ritmo en la UI y el bono de velocidad en
     *  [CardsGameContainer.finishRound]. */
    val elapsedMs: Long = 0L
  ) : CardsGameState

  data class Finished(val score: Int, val matchedPairs: Int, val attempts: Int) : CardsGameState

  data class Error(val error: Throwable) : CardsGameState
}

sealed interface CardsGameIntent : MVIIntent {
  data class CardFlipped(val id: Int) : CardsGameIntent
  data object TimerTicked : CardsGameIntent
  data object RestartGame : CardsGameIntent
}

sealed interface CardsGameAction : MVIAction {
  data class PlaySoundEffect(val effect: SoundEffect) : CardsGameAction
  data class TriggerHapticFeedback(val pattern: HapticPattern) : CardsGameAction
  data object ShowLevelUp : CardsGameAction
}

/** Ritmo de referencia para la barra de tiempo y el bono de velocidad en Precisión: 4s
 *  por pareja es un punto de partida razonable (ronda de 4 parejas "a buen ritmo" en
 *  16s), compartido entre [CardsGameContainer] (bono) y `ParejasGame.kt` (barra) para
 *  que ambos midan lo mismo. */
internal const val EXPECTED_MS_PER_PAIR = 4000L

/** Cuántos tableros sucesivos arma una sola partida antes de mostrar el resultado final.
 *  Ricardo pidió "así sucesivamente, incluso 10 niveles" -- se implementa como tope fijo
 *  (no "sin techo" como `masteryStreak`) para que Precisión, que no tiene otra forma de
 *  terminar, siga teniendo un cierre claro; si se prefiere sin límite, es cuestión de
 *  sacar el chequeo en `CardsGameContainer.finishRound`. */
internal const val TOTAL_STAGES = 10

/**
 * Cuántas parejas tiene el tablero de cada nivel (1..[TOTAL_STAGES]) -- una escalera FIJA
 * y estrictamente creciente, a pedido explícito de Ricardo tras notar que el nivel 2 le
 * había salido más fácil (menos parejas) que el nivel 1: antes la cantidad de parejas
 * salía de `VisualWorkingMemoryDDA` (D(t), que puede BAJAR si el desempeño fue malo), acá
 * en cambio depende solo del número de nivel, nunca del desempeño -- D(t) sigue afinando
 * el resto (vista previa, interferencia, distractores), pero la progresión "empieza fácil,
 * termina difícil" queda garantizada.
 */
private val PAIR_COUNTS_BY_STAGE = intArrayOf(2, 3, 4, 5, 6, 7, 8, 9, 10, 12)

internal fun pairCountForStage(stage: Int): Int =
  PAIR_COUNTS_BY_STAGE[(stage - 1).coerceIn(0, PAIR_COUNTS_BY_STAGE.lastIndex)]

/**
 * Sistema de "3 fallas" pedido por Ricardo: fallar [MISMATCHES_TO_FAIL_STAGE] parejas en
 * el tablero actual corta ese tablero ahí mismo (no espera a que se termine) y cuenta como
 * una falla de nivel. La PRIMERA falla de nivel baja un nivel; la SEGUNDA repite ese mismo
 * nivel (ya bajado) sin bajar más; la TERCERA termina la partida. Ver
 * `CardsGameContainer.handleStageFailure`.
 */
internal const val MISMATCHES_TO_FAIL_STAGE = 3
internal const val STAGE_FAILURES_BEFORE_GAME_OVER = 3

/** Bancos de símbolos graduados por similitud perceptual (0 = muy distintos entre sí,
 *  3 = máxima interferencia visual/semántica) — mismo criterio que la iteración
 *  anterior, ahora impulsado por `interferenceLevel` de [VisualWorkingMemoryDDA]. */
internal object SymbolBank {
  private val TIERS: List<List<String>> = listOf(
    listOf("🐶", "🍕", "🚗", "⭐", "🌈", "🎸", "⚽", "🎁"),
    listOf("🦊", "🐼", "🦁", "🐵", "🐺", "🐹", "🐰", "🐨", "🐸", "🐙"),
    listOf("🍎", "🍒", "🍓", "🍅", "🍇", "🍉", "🍑", "🍊"),
    listOf("😀", "😃", "😄", "😁", "😆", "🙂", "😊", "😇")
  )

  fun select(pairCount: Int, interferenceLevel: Int, random: kotlin.random.Random = kotlin.random.Random.Default): List<String> {
    val level = interferenceLevel.coerceIn(0, TIERS.lastIndex)
    val pool = mutableListOf<String>()
    for (tier in level downTo 0) {
      if (pool.size >= pairCount) break
      pool += TIERS[tier].shuffled(random)
    }
    return pool.distinct().take(pairCount)
  }
}

File: unity/NeuroVidaCore/Assets/Scripts/Games/Parejas/CardsGameContract.cs
using System.Collections.Generic;

namespace NeuroVida.Games.Parejas
{
    /// <summary>Puerto 1:1 de <c>CountdownReason</c> (Kotlin) -- distingue arranque/avance
    /// normal de una bajada o repetición de nivel por fallas, para que el cambio de nivel
    /// hacia atrás se lea como mecánica del juego, no como un bug.</summary>
    public enum CountdownReason
    {
        Start,
        Advance,
        Demoted,
        Retry
    }

    /// <summary>
    /// Puerto 1:1 de las constantes de <c>com.example.games.parejas.CardsGameContract</c>
    /// (Kotlin) -- Fase 2 del roadmap de migración. La cantidad de parejas por nivel es
    /// una escalera FIJA (nunca depende de <see cref="VisualWorkingMemoryDDA"/>, que solo
    /// afina vista previa/interferencia/distractores dentro de cada nivel) para que la
    /// progresión "empieza fácil, termina difícil" quede garantizada -- Ricardo notó
    /// jugando la versión Kotlin que el nivel 2 le había salido más fácil que el 1 cuando
    /// la cantidad de parejas todavía salía de D(t).
    /// </summary>
    public static class CardsGameContract
    {
        /// <summary>Ritmo de referencia para el bono de velocidad: 4s por pareja.</summary>
        public const long ExpectedMsPerPair = 4000L;

        /// <summary>Cuántos tableros sucesivos arma una partida antes de mostrar el
        /// resultado final.</summary>
        public const int TotalStages = 10;

        private static readonly int[] PairCountsByStage = { 2, 3, 4, 5, 6, 7, 8, 9, 10, 12 };

        public static int PairCountForStage(int stage)
        {
            int index = stage - 1;
            if (index < 0) index = 0;
            if (index > PairCountsByStage.Length - 1) index = PairCountsByStage.Length - 1;
            return PairCountsByStage[index];
        }

        /// <summary>Sistema de "3 fallas": fallar <see cref="MismatchesToFailStage"/>
        /// parejas en el tablero actual corta ese tablero ahí mismo y cuenta como una
        /// falla de nivel. 1ra falla -> baja un nivel; 2da -> repite ese mismo nivel ya
        /// bajado; 3ra -> termina la partida.</summary>
        public const int MismatchesToFailStage = 3;
        public const int StageFailuresBeforeGameOver = 3;
    }

    /// <summary>Un símbolo de carta: ícono ilustrado + variante de color (ver
    /// <see cref="SymbolSprite"/>). <see cref="Key"/> es el identificador estable usado
    /// para detectar pares (equivalente al emoji original, que servía como su propia
    /// clave por ser un string).</summary>
    public readonly struct SymbolDef
    {
        public readonly ShapeKind Shape;
        public readonly int Variant;
        public readonly string Key;

        public SymbolDef(ShapeKind shape, int variant)
        {
            Shape = shape;
            Variant = variant;
            Key = $"{shape}:{variant}";
        }
    }

    /// <summary>Puerto de <c>SymbolBank</c> (Kotlin) -- bancos de símbolos graduados por
    /// similitud perceptual (0 = muy distintos entre sí, 3 = máxima interferencia),
    /// impulsado por <c>DifficultyProfile.InterferenceLevel</c>.
    ///
    /// La versión Kotlin usa emojis de color (el font system de Android los renderiza
    /// nativo). Acá son íconos ilustrados procedurales (<see cref="SymbolSprite"/>): la
    /// fuente legacy de Unity (<c>LegacyRuntime.ttf</c>) no tiene glifos de emoji, así que
    /// las cartas se veían vacías (bug reportado por Ricardo jugando en el dispositivo,
    /// 23-sep).
    ///
    /// Gradiente de dificultad, siempre con colores vívidos (Ricardo: "los símbolos son
    /// muy opacos, oscuros, difíciles de recordar"): tier 0 = objetos y colores muy
    /// distintos; tier 1 = más variedad de objetos; tier 2 = mismo objeto en dos colores
    /// distintos (hay que recordar objeto Y color); tier 3 = mismo objeto en dos tonos
    /// análogos (variantes 0 y 3 de cada ícono).</summary>
    public static class SymbolBank
    {
        private static readonly SymbolDef[][] Tiers =
        {
            // Tier 0: 8 objetos cotidianos, cada uno de un color dominante distinto.
            new[]
            {
                new SymbolDef(ShapeKind.Apple, 0),
                new SymbolDef(ShapeKind.Star, 0),
                new SymbolDef(ShapeKind.Drop, 0),
                new SymbolDef(ShapeKind.Sun, 0),
                new SymbolDef(ShapeKind.Flower, 0),
                new SymbolDef(ShapeKind.Clover, 0),
                new SymbolDef(ShapeKind.Cloud, 0),
                new SymbolDef(ShapeKind.Crown, 0),
            },
            // Tier 1: más objetos, algunos repiten forma con otro color.
            new[]
            {
                new SymbolDef(ShapeKind.Heart, 0),
                new SymbolDef(ShapeKind.Fish, 0),
                new SymbolDef(ShapeKind.Mushroom, 0),
                new SymbolDef(ShapeKind.Balloon, 2),
                new SymbolDef(ShapeKind.Moon, 1),
                new SymbolDef(ShapeKind.Gem, 0),
                new SymbolDef(ShapeKind.Apple, 1),
                new SymbolDef(ShapeKind.Star, 1),
                new SymbolDef(ShapeKind.Flower, 1),
                new SymbolDef(ShapeKind.Drop, 1),
            },
            // Tier 2: pares del mismo objeto en dos colores claramente distintos.
            new[]
            {
                new SymbolDef(ShapeKind.Heart, 0),
                new SymbolDef(ShapeKind.Heart, 1),
                new SymbolDef(ShapeKind.Balloon, 0),
                new SymbolDef(ShapeKind.Balloon, 1),
                new SymbolDef(ShapeKind.Fish, 0),
                new SymbolDef(ShapeKind.Fish, 2),
                new SymbolDef(ShapeKind.Crown, 0),
                new SymbolDef(ShapeKind.Crown, 1),
            },
            // Tier 3 (máxima interferencia): pares del mismo objeto en tonos análogos.
            new[]
            {
                new SymbolDef(ShapeKind.Apple, 0),
                new SymbolDef(ShapeKind.Apple, 3),
                new SymbolDef(ShapeKind.Star, 0),
                new SymbolDef(ShapeKind.Star, 3),
                new SymbolDef(ShapeKind.Drop, 0),
                new SymbolDef(ShapeKind.Drop, 3),
                new SymbolDef(ShapeKind.Flower, 0),
                new SymbolDef(ShapeKind.Flower, 3),
            },
        };

        public static List<SymbolDef> Select(int pairCount, int interferenceLevel, System.Random random)
        {
            int level = interferenceLevel;
            if (level < 0) level = 0;
            if (level > Tiers.Length - 1) level = Tiers.Length - 1;

            var pool = new List<SymbolDef>();
            var seenKeys = new HashSet<string>();
            void AddFromTier(int tier)
            {
                foreach (var symbol in Shuffled(Tiers[tier], random))
                {
                    if (pool.Count >= pairCount) return;
                    if (!seenKeys.Add(symbol.Key)) continue;
                    pool.Add(symbol);
                }
            }

            for (int tier = level; tier >= 0 && pool.Count < pairCount; tier--) AddFromTier(tier);
            // La escalera fija de niveles (CardsGameContract.PairCountForStage) llega a
            // pairCount=12, pero con interferenceLevel bajo (0 o 1) los tiers 0..level solos
            // no siempre alcanzan esa cantidad (tier 0 solo tiene 8 símbolos distintos).
            // Antes esto dejaba el mazo con MENOS pares de los que _totalPairs esperaba --
            // el tablero nunca podía completarse (partida "pegada", Ricardo lo vio cerca del
            // nivel 8) y sobraban casillas de grilla sin carta asignada. Completar con los
            // tiers restantes (aunque sean "más difíciles") garantiza siempre pairCount
            // símbolos -- más de 25 distintos entre los 4 tiers, contra un máximo de 12.
            for (int tier = level + 1; tier < Tiers.Length && pool.Count < pairCount; tier++) AddFromTier(tier);

            return pool;
        }

        private static IEnumerable<SymbolDef> Shuffled(SymbolDef[] source, System.Random random)
        {
            var list = new List<SymbolDef>(source);
            for (int i = list.Count - 1; i > 0; i--)
            {
                int j = random.Next(i + 1);
                (list[i], list[j]) = (list[j], list[i]);
            }
            return list;
        }
    }
}
